Core/clim-basic/input.lisp

```diff
@@ -288,79 +298,71 @@ use condition-variables nor locks."))
-    (process-next-event port :timeout 0)
-    (loop
-       as decay = (compute-decay timeout-time (event-schedule-time queue))
-       do
-         (when decay (maxf decay 0))
-         (check-schedule queue)
-         (cond ((event-queue-head queue)
-                (return t))
-               ((maybe-funcall wait-function)
-                (return (values nil :wait-function)))
-               ((and timeout-time (> (now) timeout-time))
-                (return (values nil :timeout)))
-               (T
-                (process-next-event port :timeout decay :wait-function wait-function))))))
+  (check-schedule queue)
+  (process-next-event (event-queue-port queue) :timeout 0)
+  (loop
+     with port = (event-queue-port queue)
+     with timeout-time = (and timeout (+ timeout (now)))
+     as decay = (compute-decay timeout-time (event-schedule-time queue))
+     do
+       (cond ((event-queue-head queue)
+              (return t))
+             ((and timeout-time (> (now) timeout-time))
+              (return (values nil :timeout)))
+             ((multiple-value-bind (available reason)
+                  (process-next-event port :timeout decay :wait-function wait-function)
+                (and (null available)
+                     (eql reason :wait-function)))
+              (return (values nil :wait-function)))
+             (t (check-schedule queue)))))
 
 (defmethod event-queue-listen ((queue simple-event-queue))
   (event-queue-listen-or-wait queue :timeout 0))
 
 
+;;; XXX SBCL doesn't reacquire lock when condition-variable
+;;; prematurely returns (i.e due to timeout). --jd 2019-06-26
+
 (defclass concurrent-event-queue (simple-event-queue)
-  ((lock :initform (make-recursive-lock "event queue")
+  ((lock :initform (make-lock "Event queue")
          :reader event-queue-lock)
    (processes
-         :initform (make-condition-variable)
-         :accessor event-queue-processes
-         :documentation "Condition variable for waiting processes")))
-
-(defmethod check-schedule :around ((queue concurrent-event-queue))
-  (with-recursive-lock-held ((event-queue-lock queue))
-    (call-next-method)))
-
-(defmethod schedule-event-queue :around ((queue concurrent-event-queue) event delay)
-  (with-recursive-lock-held ((event-queue-lock queue))
-    (call-next-method)))
+    :initform (make-condition-variable)
+    :accessor event-queue-processes
+    :documentation "Condition variable for waiting processes")))
 
 (defmethod event-queue-read ((queue concurrent-event-queue))
   (do-port-force-output queue)
   (let ((lock (event-queue-lock queue))
         (cv (event-queue-processes queue)))
     (loop
        as decay = (compute-decay nil (event-schedule-time queue))
-       do (with-recursive-lock-held (lock)
-            (if-let ((result (%event-queue-read queue)))
-              (return result)
-              (condition-wait cv lock decay))))))
+       do
+         (check-schedule queue)
```

Comment:
because that would cause a deadlock (that's the main reason why the initial implementation had recursive locks).
check-schedule may call event-append which has a critical section on its own, so this call outside the critical section is safe.

One could wonder, if access to the scheduler-mixin slots is safe. It is, because there is only one consumer (read/peek/listen etc.) who calls check-schedule - i.e producent doesn't call it. We always consume queue from the CLIM loop (and it is not exported API except the method `schedule-event` which operates on sheets and these operations are also called from CLIM loop.